File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
import heapq


class MedianFinder:
    def __init__(self):
        self.one = list()
        self.two = list()


    def addNum(self, num: int) -> None:
        if len(self.one) == len(self.two):
            heapq.heappush(self.one, (-num, num))
        else:
            heapq.heappush(self.two, (num, num))
        if self.two and self.two[0][1] < self.one[0][1]:
            n1 = heapq.heappop(self.one)[1]
            n2 = heapq.heappop(self.two)[1]
            heapq.heappush(self.one, (-n2, n2))
            heapq.heappush(self.two, (n1, n1))


    def findMedian(self) -> float:
        if len(self.one) == len(self.two):
            return (self.one[0][1] + self.two[0][1]) / 2
        return self.one[0][1]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.java
class MedianFinder {
    private PriorityQueue<Integer> left;
    private PriorityQueue<Integer> right;

    public MedianFinder() {
        left = new PriorityQueue<>();
        right = new PriorityQueue<>();
    }

    public void addNum(int num) {
        if (left.size() != right.size()) {
            left.add(-1 * num);
        } else {
            right.add(num);
        }
        if (!left.isEmpty() && right.peek() < -1 * left.peek()) {
            int n1 = -1 * left.poll();
            int n2 = right.poll();
            left.add(-1 * n2);
            right.add(n1);
        }
    }

    public double findMedian() {
        double result;

        if (left.size() == right.size()) {
            result = (-1 * left.peek() + (double)right.peek()) / 2;
        } else {
            result = right.peek();
        }
        return result;
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */